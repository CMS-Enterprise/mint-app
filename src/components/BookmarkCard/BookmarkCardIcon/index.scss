.bookmark {
    &__icon {
        color: #005EA2;
        display: inline-block;
        cursor: pointer;
    }
}

.black {
    color: black;
}

.lightgrey {
    color: $medium-gray
}
