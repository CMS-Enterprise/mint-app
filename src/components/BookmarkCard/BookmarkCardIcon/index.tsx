import React from 'react';
import classnames from 'classnames';

import './index.scss';

type BookmarkCardIconProps = {
  className?: string;
  color?: 'black' | 'lightgrey';
  size: 'sm' | 'md' | 'lg';
  onClick?: React.MouseEventHandler<HTMLButtonElement>;
};

const BookmarkCardIcon = ({
  size,
  color,
  className,
  onClick
}: BookmarkCardIconProps) => {
  const classes = classnames(
    'fa',
    'fa-bookmark',
    'bookmark__icon',
    {
      'fa-1x': size === 'sm'
    },
    {
      'fa-2x': size === 'md'
    },
    {
      'fa-3x': size === 'lg'
    },
    {
      lightgrey: color === 'lightgrey'
    },
    {
      black: color === 'black'
    },
    className
  );

  return (
    <i
      className={classes}
      onClick={onClick}
      data-testid="bookmark-icon"
      aria-hidden="true"
    />
  );
};

export default BookmarkCardIcon;
