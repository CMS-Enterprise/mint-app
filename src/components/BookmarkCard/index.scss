.bookmark {
    display: flex;
    flex-direction: column;
    align-items: flex-start;

    &__container {        
        padding: 24px;
        height: 350px;
        border: 2px solid #E6E6E6;
        border-radius: 4px;
    }

    &__header {
        width: 100%;
    }

    &__title {
        overflow: hidden;
        white-space: nowrap;
        text-overflow: ellipsis;
        max-width: 90%;
    }

    &__body-text {
        display: -webkit-box;
        -webkit-line-clamp: 3;
        -webkit-box-orient: vertical;
        overflow: hidden;
        min-height: 4.5em;
    }
}

#main-content .usa-card {
    margin-bottom: 1em;
}
