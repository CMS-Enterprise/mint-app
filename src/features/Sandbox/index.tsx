import React, { useEffect } from 'react';
import { GridContainer, Icon } from '@trussworks/react-uswds';

import MainContent from 'components/MainContent';

import NestedTable from './DnD/NestedTable';

export type ColumnSortType = {
  isSorted: boolean;
  isSortedDesc: boolean;
};

export type MilestoneType = {
  id: string;
  risk: string;
  name: string;
  facilitatedBy: string;
  solutions: string[];
  needBy: string;
  status: string;
  actions: any;
};

export type SubCategoryType = {
  id: string;
  risk: string;
  name: string;
  facilitatedBy: string;
  solutions: string[];
  needBy: string;
  status: string;
  actions: any;
  milestones: MilestoneType[];
};

export type CategoryType = {
  id: string;
  risk: string;
  name: string;
  facilitatedBy: string;
  solutions: string[];
  needBy: string;
  status: string;
  actions: any;
  subCategories: SubCategoryType[];
};

const rawData: CategoryType[] = [
  {
    id: '1',
    risk: '',
    name: 'Category 1',
    facilitatedBy: '',
    solutions: [],
    needBy: '',
    status: '',
    actions: '',
    subCategories: [
      {
        id: '1-1',
        risk: '',
        name: 'Sub-Category 1',
        facilitatedBy: '',
        solutions: [],
        needBy: '',
        status: '',
        actions: '',
        milestones: [
          {
            id: '1-1-1',
            risk: 'H',
            name: 'Milestone 1',
            facilitatedBy: 'Facilitator 1',
            solutions: ['Solution 1', 'Solution 2'],
            needBy: '2022-01-01',
            status: 'In Progress',
            actions: 'Actions 1'
          },
          {
            id: '1-1-2',
            risk: 'L',
            name: 'Milestone 2',
            facilitatedBy: 'Facilitator 2',
            solutions: ['Solution 3', 'Solution 4'],
            needBy: '2022-01-02',
            status: 'In Progress',
            actions: 'Actions 2'
          }
        ]
      },
      {
        id: '1-2',
        risk: '',
        name: 'Sub-Category 2',
        facilitatedBy: '',
        solutions: [],
        needBy: '',
        status: '',
        actions: '',
        milestones: [
          {
            id: '1-2-1',
            risk: 'M',
            name: 'Milestone 3',
            facilitatedBy: 'Facilitator 3',
            solutions: ['Solution 5', 'Solution 6'],
            needBy: '2022-01-03',
            status: 'In Progress',
            actions: 'Actions 3'
          },
          {
            id: '1-2-2',
            risk: 'H',
            name: 'Milestone 4',
            facilitatedBy: 'Facilitator 4',
            solutions: ['Solution 7', 'Solution 8'],
            needBy: '2022-01-04',
            status: 'In Progress',
            actions: 'Actions 4'
          }
        ]
      }
    ]
  },
  {
    id: '2',
    risk: '',
    name: 'Category 2',
    facilitatedBy: '',
    solutions: [],
    needBy: '',
    status: '',
    actions: '',
    subCategories: [
      {
        id: '2-1',
        risk: '',
        name: 'Sub-Category 3',
        facilitatedBy: '',
        solutions: [],
        needBy: '',
        status: '',
        actions: '',
        milestones: [
          {
            id: '2-1-1',
            risk: 'L',
            name: 'Milestone 5',
            facilitatedBy: 'Facilitator 1',
            solutions: ['Solution 1', 'Solution 2'],
            needBy: '2022-01-01',
            status: 'In Progress',
            actions: 'Actions 1'
          },
          {
            id: '2-1-2',
            risk: 'H',
            name: 'Milestone 6',
            facilitatedBy: 'Facilitator 2',
            solutions: ['Solution 3', 'Solution 4'],
            needBy: '2022-01-02',
            status: 'In Progress',
            actions: 'Actions 2'
          }
        ]
      },
      {
        id: '2-2',
        risk: '',
        name: 'Sub-Category 4',
        facilitatedBy: '',
        solutions: [],
        needBy: '',
        status: '',
        actions: '',
        milestones: [
          {
            id: '2-2-1',
            risk: 'M',
            name: 'Milestone 7',
            facilitatedBy: 'Facilitator 3',
            solutions: ['Solution 5', 'Solution 6'],
            needBy: '2022-01-03',
            status: 'In Progress',
            actions: 'Actions 3'
          },
          {
            id: '2-2-2',
            risk: 'H',
            name: 'Milestone 8',
            facilitatedBy: 'Facilitator 4',
            solutions: ['Solution 7', 'Solution 8'],
            needBy: '2022-01-04',
            status: 'In Progress',
            actions: 'Actions 4'
          }
        ]
      }
    ]
  }
];

const riskMap: { [key: string]: number } = {
  Low: 1,
  Medium: 2,
  High: 3
};

type ColumnType = {
  Header: string | React.ReactNode;
  accessor: string;
  width: string;
  canSort?: boolean;
  sort?: (
    data: CategoryType[],
    direction: 'ASC' | 'DESC',
    accessor: keyof MilestoneType
  ) => CategoryType[];
};

const sortNested = (
  data: CategoryType[],
  direction: 'ASC' | 'DESC',
  accessor: keyof MilestoneType
) => {
  data.forEach(category => {
    category.subCategories.forEach(subCategory => {
      subCategory.milestones.sort((a, b) =>
        a[accessor].localeCompare(b[accessor])
      );
      if (direction === 'ASC') {
        subCategory.milestones.reverse();
      }
    });
  });
  return data;
};

export const columns: ColumnType[] = [
  {
    Header: <Icon.Warning size={3} className="left-05 text-base-lighter" />,
    accessor: 'risk',
    width: '60px',
    sort: (
      data: CategoryType[],
      direction: 'ASC' | 'DESC',
      accessor: keyof MilestoneType
    ) => {
      data.forEach(category => {
        category.subCategories.forEach(subCategory => {
          subCategory.milestones.sort(
            (a, b) => (riskMap[a.risk] || 0) - (riskMap[b.risk] || 0)
          );
          if (direction === 'ASC') {
            subCategory.milestones.reverse();
          }
        });
      });
      return data;
    }
  },
  {
    Header: 'Model milestone',
    accessor: 'name',
    width: '200px',
    sort: sortNested
  },
  {
    Header: 'Facilitated By',
    accessor: 'facilitatedBy',
    width: '200px',
    sort: sortNested
  },
  {
    Header: 'Solutions',
    accessor: 'solutions',
    width: '200px',
    sort: (
      data: CategoryType[],
      direction: 'ASC' | 'DESC',
      accessor: keyof MilestoneType
    ) => {
      data.forEach(category => {
        category.subCategories.forEach(subCategory => {
          subCategory.milestones.sort((a, b) =>
            a.solutions.join().localeCompare(b.solutions.join())
          );
          if (direction === 'ASC') {
            subCategory.milestones.reverse();
          }
        });
      });
      return data;
    }
  },
  {
    Header: 'Need By',
    accessor: 'needBy',
    width: '130px',
    sort: sortNested
  },
  {
    Header: 'Status',
    accessor: 'status',
    width: '130px',
    sort: sortNested
  },
  {
    Header: 'Actions',
    accessor: 'actions',
    width: '130px',
    canSort: false
  }
];

const Sandbox = () => {
  useEffect(() => {
    document.title = 'Sandbox';
  }, []);

  return (
    <MainContent>
      <GridContainer className="margin-y-6">
        <NestedTable rawData={rawData} />
      </GridContainer>
    </MainContent>
  );
};

export default Sandbox;
