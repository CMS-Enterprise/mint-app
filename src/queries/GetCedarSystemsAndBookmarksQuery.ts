import { gql } from '@apollo/client';

export default gql`
  query GetCedarSystemsAndBookmarks {
    cedarSystemBookmarks {
      euaUserId
      cedarSystemId
    }
    cedarSystems {
      id
      name
      description
      acronym
      status
      businessOwnerOrg
      businessOwnerOrgComp
      systemMaintainerOrg
      systemMaintainerOrgComp
    }
  }
`;
