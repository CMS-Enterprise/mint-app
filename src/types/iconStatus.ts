export type IconStatus = 'success' | 'warning' | 'fail'; // TODO - Need more types for IconStatus in response to CEDAR SystemHealthIcon mappings.

export type CedarStatus = string | null; // Need to extract all possible returns for this field from CEDAR ex: Approved, Draft, etc.

// TODO - if we want to keep this text past the prototype state, it needs to use translation
export const mapCedarStatusToIcon = (status: CedarStatus): IconStatus => {
  switch (status) {
    case 'Approved':
      return 'success';
    case 'Draft':
      return 'warning';
    default:
      return 'fail';
  }
};

const mapStatusToRanking = (status: IconStatus): number => {
  switch (status) {
    case 'success':
      return 0;
    case 'warning':
      return 1;
    case 'fail':
      return 2;
    default:
      return Number.MAX_SAFE_INTEGER;
  }
};

export const sortByStatus = (
  statusA: IconStatus,
  statusB: IconStatus
): number => {
  return mapStatusToRanking(statusA) - mapStatusToRanking(statusB);
};
