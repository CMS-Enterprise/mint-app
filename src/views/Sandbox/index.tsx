import React, { useEffect } from 'react';

import MainContent from 'components/MainContent';

const Sandbox = () => {
  useEffect(() => {
    document.title = 'Sandbox';
  }, []);

  return (
    <MainContent>
      <></>
    </MainContent>
  );
};

export default Sandbox;
