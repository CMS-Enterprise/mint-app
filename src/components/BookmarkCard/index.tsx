import React from 'react';
import { useTranslation } from 'react-i18next';
import { useMutation } from '@apollo/client';
import { Card } from '@trussworks/react-uswds';
import classnames from 'classnames';

import UswdsReactLink from 'components/LinkWrapper';
import Divider from 'components/shared/Divider';
import SystemHealthIcon from 'components/SystemHealthIcon';
import DeleteCedarSystemBookmarkQuery from 'queries/DeleteCedarSystemBookmarkQuery';
import { GetCedarSystems_cedarSystems as CedarSystemProps } from 'queries/types/GetCedarSystems';
import { IconStatus } from 'types/iconStatus';

import BookmarkCardIcon from './BookmarkCardIcon';

import './index.scss';

type BookmarkCardProps = {
  className?: string;
  statusIcon: IconStatus;
  type: 'systemProfile'; // Built in for future iterations/varations of bookmarked datasets that ingest i18n translations for headers.
  refetch: () => any | undefined;
};

const BookmarkCard = ({
  className,
  type,
  id,
  name,
  description,
  acronym,
  status,
  statusIcon,
  businessOwnerOrg,
  refetch
}: BookmarkCardProps & CedarSystemProps) => {
  const { t } = useTranslation();

  const [deleteMutate] = useMutation(DeleteCedarSystemBookmarkQuery);

  const handleDeleteBookmark = (cedarSystemId: string) => {
    deleteMutate({
      variables: {
        input: {
          cedarSystemId
        }
      }
    }).then(refetch);
  };

  return (
    <Card
      data-testid="single-bookmark-card"
      className={classnames('desktop:grid-col-6', 'grid-col-12', className)}
    >
      <div className="grid-col-12">
        <div className="bookmark__header easi-header__basic">
          <h2 className="bookmark__title margin-top-0 margin-bottom-1">
            <UswdsReactLink to={`/systems/${id}/home/top`}>
              {name}
            </UswdsReactLink>
          </h2>
          <BookmarkCardIcon
            size="md"
            onClick={() => handleDeleteBookmark(id)}
          />
        </div>
        <p className="margin-0">{acronym}</p>
        <p className="bookmark__body-text line-height-body-4">{description}</p>
        <p className="margin-bottom-0">{t(`${type}:bookmark.subHeader1`)}</p>
        <p className="text-bold margin-top-1">{businessOwnerOrg}</p>
        <Divider />
        <div className="bookmark__header easi-header__basic">
          <div>
            <p className="margin-bottom-0">
              {t(`${type}:bookmark.subHeader2`)}
            </p>
            <p className="text-bold margin-top-0 margin-bottom-0">{status}</p>
          </div>
          <SystemHealthIcon
            status={statusIcon}
            size="xl"
            className="margin-top-3"
          />
        </div>
      </div>
    </Card>
  );
};

export default BookmarkCard;
