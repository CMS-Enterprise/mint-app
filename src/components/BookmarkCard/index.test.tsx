import React from 'react';
import { MemoryRouter } from 'react-router-dom';
import { MockedProvider } from '@apollo/client/testing';
import { render } from '@testing-library/react';

import GetCedarSystemsAndBookmarksQuery from 'queries/GetCedarSystemsAndBookmarksQuery';
import { mapCedarStatusToIcon } from 'types/iconStatus';
import { mockSystemInfo } from 'views/Sandbox/mockSystemData';

import BookmarkCard from './index';

describe('BookmarkCard', () => {
  const mocks = [
    {
      request: {
        query: GetCedarSystemsAndBookmarksQuery
      },
      result: {
        data: {
          cedarSystems: [],
          cedarSystemBookmarks: []
        }
      }
    }
  ];

  it('matches the snapshot', () => {
    const { asFragment } = render(
      <MemoryRouter>
        <MockedProvider mocks={mocks}>
          <BookmarkCard
            type="systemProfile"
            statusIcon={mapCedarStatusToIcon(mockSystemInfo[0].status)}
            refetch={() => {}}
            {...mockSystemInfo[0]}
          />
        </MockedProvider>
      </MemoryRouter>
    );
    expect(asFragment()).toMatchSnapshot();
  });

  it('renders translated headings', () => {
    const { getByText } = render(
      <MemoryRouter>
        <MockedProvider mocks={mocks}>
          <BookmarkCard
            type="systemProfile"
            statusIcon={mapCedarStatusToIcon(mockSystemInfo[0].status)}
            refetch={() => {}}
            {...mockSystemInfo[0]}
          />
        </MockedProvider>
      </MemoryRouter>
    );

    // TODO Update expected text output when translations/headings of systemList get solidifed
    expect(getByText('Happiness Achievement Module')).toBeInTheDocument();
    expect(getByText('CMS Component')).toBeInTheDocument();
    expect(getByText('ATO Status')).toBeInTheDocument();
  });

  it('renders corresponding success health icon for status', () => {
    const { getByTestId } = render(
      <MemoryRouter>
        <MockedProvider mocks={mocks}>
          <BookmarkCard
            type="systemProfile"
            statusIcon={mapCedarStatusToIcon(mockSystemInfo[0].status)}
            refetch={() => {}}
            {...mockSystemInfo[0]}
          />
        </MockedProvider>
      </MemoryRouter>
    );

    expect(getByTestId('system-health-icon')).toHaveClass(
      'system-health-icon-success'
    );
  });

  it('renders corresponding warning health icon for status', () => {
    const { getByTestId } = render(
      <MemoryRouter>
        <MockedProvider mocks={mocks}>
          <BookmarkCard
            type="systemProfile"
            statusIcon={mapCedarStatusToIcon(mockSystemInfo[1].status)}
            refetch={() => {}}
            {...mockSystemInfo[1]}
          />
        </MockedProvider>
      </MemoryRouter>
    );

    expect(getByTestId('system-health-icon')).toHaveClass(
      'system-health-icon-fail'
    );
  });

  it('renders corresponding fail health icon for status', () => {
    const { getByTestId } = render(
      <MemoryRouter>
        <MockedProvider mocks={mocks}>
          <BookmarkCard
            type="systemProfile"
            statusIcon={mapCedarStatusToIcon(mockSystemInfo[2].status)}
            refetch={() => {}}
            {...mockSystemInfo[2]}
          />
        </MockedProvider>
      </MemoryRouter>
    );

    expect(getByTestId('system-health-icon')).toHaveClass(
      'system-health-icon-warning'
    );
  });
});
