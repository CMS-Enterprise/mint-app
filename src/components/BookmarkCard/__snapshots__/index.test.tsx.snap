// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BookmarkCard matches the snapshot 1`] = `
<DocumentFragment>
  <li
    class="usa-card desktop:grid-col-6 grid-col-12"
    data-testid="single-bookmark-card"
  >
    <div
      class="usa-card__container"
    >
      <div
        class="grid-col-12"
      >
        <div
          class="bookmark__header easi-header__basic"
        >
          <h2
            class="bookmark__title margin-top-0 margin-bottom-1"
          >
            <a
              class="usa-link"
              href="/systems/1/home/top"
            >
              Happiness Achievement Module
            </a>
          </h2>
          <i
            aria-hidden="true"
            class="fa fa-bookmark bookmark__icon fa-2x"
            data-testid="bookmark-icon"
          />
        </div>
        <p
          class="margin-0"
        >
          HAM
        </p>
        <p
          class="bookmark__body-text line-height-body-4"
        >
          Lorem ipsum dolor sit amet, consectetur adipiscing elit.
    Sollicitudin donec aliquam dui sed odio porta. Faucibus quam egestas
    feugiat laoreet quis. Sapien, sagittis, consectetur adipiscing elit.
    Sollicitudin donec aliquam dui sed odio porta.
        </p>
        <p
          class="margin-bottom-0"
        >
          CMS Component
        </p>
        <p
          class="text-bold margin-top-1"
        >
          CMMI
        </p>
        <div
          class="easi-divider"
        />
        <div
          class="bookmark__header easi-header__basic"
        >
          <div>
            <p
              class="margin-bottom-0"
            >
              ATO Status
            </p>
            <p
              class="text-bold margin-top-0 margin-bottom-0"
            >
              Approved
            </p>
          </div>
          <i
            class="fa system-health-icon-success fa-check-circle fa-3x margin-top-3"
            data-testid="system-health-icon"
          />
        </div>
      </div>
    </div>
  </li>
</DocumentFragment>
`;
